### PR TITLE
Vulnerability restapi

### PR DESCRIPTION
**Pull request by cg122**
## Description
Added new endpoints for vulnerability  information.

## Motivation and context
Since new architecture for processing and storing vulnerability related information is deployed, it is now feasible to get vulnerability information in a more straightforward manner.

## Task list  
- [x] Implement endpoints
- [x] Local testing the endpoints

### 1. Get all vulnerabilities

`GET /mvn/vulnerabilities/`

Query Params: 
   - limit: integer, default=10
   - offset: integer, default=0
   - attributes: list:text, default=""

REPSONSE CODE:

RESPONSE BODY: 

```
[ {
   "id": BIGINT,
   "external_id": TEXT,
   "statement": JSON //if attributes provided, includes only listed fields
} ]
```

### 2. Find vulnerability by external id (source specific id, e.g. CVE-2018-14041, or GHSA-2pwh-52h7-7j84 )
`GET /mvn/vulnerabilities/{external_id}`

Query Params: 
   - attributes: list:text, default=""

REPSONSE CODE:

RESPONSE BODY:

```
{
   "id": BIGINT,
   "external_id": TEXT,
   "statement": JSON //if attributes provided, includes only listed fields
}
```

### 3. Get purls mapped to a vulnerability of the external id

`GET /mvn/vulnerabilities/{external_id}/purls`

Query Params: 
   - limit: integer, default=10
   - offset: integer, default=0

REPSONSE CODE:

RESPONSE BODY:

```
[{
   "vulnerability_id": BIGINT
   "purl": TEXT,
   "package_name": TEXT,
   "package_version": TEXT,
   "forge": TEXT
}]
```

### 4. Get callables mapped to a vulnerability of the external id

`GET /mvn/vulnerabilities/{external_id}/callables`

Params: 
   - limit: integer, default=10
   - offset: integer, default=0

REPSONSE CODE:
RESPONSE BODY:

```
[{
   "fasten_uri": TEXT,
   "callable_id": BIGINT
}]
```
### 5. Get vulnerabilities of given maven coordinates

`POST /mvn/vulnerabilities/coordinates/`

Query Params: 
   - attributes: list:text, default=""

REQUEST BODY: 
   - List of Maven coordinates (better to check if coordinates are in the KB with the endpoint `mvn/packages/{pkg}/{pkg_ver}` before invoke, then it could trigger lazy injection if coordinate is not there yet)

REPSONSE CODE:

RESPONSE BODY:

```
[{
   "$coordinate":"[{
                      "id": BIGINT,
                      "external_id": TEXT,
                      "statement": JSON //if attributes provided, includes only listed fields
                   }]"
}]
```

### 6. Get vulnerabilities of a package version (aka. maven coordinate, invoking this endpoint may trigger lazy injection of package version. Will it also trigger the vulnerability to update?)

`GET /mvn/packages/{pkg}/{pkg_version}/vulnerabilities`

Query Params: 
   - limit: integer, default=10
   - offset: integer, default=0
   - attributes: list:text, default=""

REPSONSE CODE:
200: Success.
201: Package version is not ready, but being created now.
400: Package version not found.

RESPONSE BODY:

```
[{
   "id": BIGINT,
   "external_id": TEXT,
   "statement": JSON //if attributes provided, includes only listed fields
}]
```

### CODE REVIEW

**Comment by MagielBruntink** [comment]
I tested all GET endpoints in a local docker-compose deployment and it works fine. Great!

**Comment by mir-am** [comment]
I will take a look at the PR by the end of the week.

**Comment by proksch** [comment]
@mir-am I would appreciate if you could review the changes timely, so we can merge them into develop.

File: .github/workflows/dockerpublish_restapi.yaml [hunk 1]
```diff
@@ -2,18 +2,18 @@ name: Publish Docker image
 
 on:
   push:
-    # Publish `develop` as Docker `latest` image.
+    # Publish `vulnerability-restapi` as Docker `latest` image.
     branches:
-      - develop
+      - vulnerability-restapi
```
Comment (mir-am) [inline .github/workflows/dockerpublish_restapi.yaml:7]: Why the branch name is changed here? I guess this is done for testing? right?

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/PackageApi.java [hunk 2]
```diff
@@ -93,5 +93,7 @@ public PackageApi(PackageApiService service) {
                                        @RequestParam(required = false) Long releaseDate) {
         return service.getERCGLink(packageName, version, artifactRepo, releaseDate);
     }
+
```
Comment (mir-am) [inline analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/PackageApi.java:96]: Could you please remove these two blank lines?

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/impl/VulnerabilityApiServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        var statements = KnowledgeBaseConnector.kbDao.getPackageVersionVulnerabilities(package_name, package_version, true);
+        JSONArray vulnerabilities = new JSONArray(statements);
+        var result = getStatementsAttributes(vulnerabilities, attributes)
+                .toString().replace("\\/", "/");
+        return new ResponseEntity<>(result, HttpStatus.OK);
+    }
+
+    @Override
+    public ResponseEntity<String> getCallableVulnerability(String package_name, String package_version, String fasten_uri) {
+        return null;
```
Comment (mir-am) [inline analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/impl/VulnerabilityApiServiceImpl.java:118]: Why does this method return null?

**Review by mir-am — COMMENTED** [review]
Overall, LGTM.
I have made some minor comments.

**Comment by mir-am** [comment]
@MagielBruntink, @cg122
I very much appreciate it if you could also update the two wiki pages below after the recent updates to the metadata DB schema and also the introduction of new endpoints.
https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects
https://github.com/fasten-project/fasten/wiki/Metadata-Database-Schema

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/impl/VulnerabilityApiServiceImpl.java (hunk 3 above)
Comment (cg122) [reply analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/impl/VulnerabilityApiServiceImpl.java:118]: Removed, not implemented now.

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/PackageApi.java (hunk 2 above)
Comment (cg122) [reply analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/mvn/api/PackageApi.java:96]: Done

File: .github/workflows/dockerpublish_restapi.yaml (hunk 1 above)
Comment (cg122) [reply .github/workflows/dockerpublish_restapi.yaml:7]: Yes, for testing purpose. Changed back now.

**Comment by cg122** [comment]
> 

Updated related pages, except for the schema diagram.

**Comment by MagielBruntink** [comment]
This seems ready to be merged, agree @mir-am

**Comment by proksch** [comment]
Looks good to me. Would it be possible to rebase the branch on develop instead of merging (and get rid of the merge commits)?

**Comment by mir-am** [comment]
> https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects

LGTM! Please merge the PR.
My only question is what are the required steps to deploy this and also the new vulnerability architecture on the FASTEN clusters (e.g. monster)? I assume we also need to create DB tables.

**Comment by MagielBruntink** [comment]
Merging the old way, as rebasing the existing commits locally leads to issues here.

**Comment by MagielBruntink** [comment]
> > https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects
> 
> LGTM! Please merge the PR. My only question is what are the required steps to deploy this and also the new vulnerability architecture on the FASTEN clusters (e.g. monster)? I assume we also need to create DB tables.

The DB table & index creation is documented here:
https://github.com/fasten-project/fasten-docker-deployment/blob/162704c2a7d77aee9433ab6db903f7907d6b8b9a/metadata-db/psql_create_tables_java.sql#L230
https://github.com/fasten-project/fasten-docker-deployment/blob/162704c2a7d77aee9433ab6db903f7907d6b8b9a/metadata-db/psql_create_tables_c.sql#L228
https://github.com/fasten-project/fasten-docker-deployment/blob/162704c2a7d77aee9433ab6db903f7907d6b8b9a/metadata-db/psql_create_tables_python.sql#L230

To get data into the new tables and make the new part of the REST API useful, the vulnerability-statements-processor and vulnerability-packages-listener need to be deployed. Then you can shut down the old vulnerability-consumer as well.

Deployment configuration for these two plugins is here:
statements-processor: https://github.com/fasten-project/fasten-docker-deployment/blob/162704c2a7d77aee9433ab6db903f7907d6b8b9a/docker-compose.yml#L239
packages-listener: https://github.com/fasten-project/fasten-docker-deployment/blob/162704c2a7d77aee9433ab6db903f7907d6b8b9a/docker-compose.yml#L268

**Pull request merged**